jax/_src/nn/functions.py

```diff
@@ -765,3 +767,172 @@ def hard_silu(x: ArrayLike) -> Array:
+  padded_logits = padded_logits.astype(jnp.float32)
+  probs = jax.nn.softmax(padded_logits, axis=-1).astype(key.dtype)
+
+  # Compute the attention context.
+  encoded = jnp.einsum('BNTS,BSNH->BTNH', probs, value)
+  return encoded
+
+def dot_product_attention(
+    query: ArrayLike,
+    key: ArrayLike,
+    value: ArrayLike,
+    bias: ArrayLike | None = None,
+    mask: ArrayLike | None = None,
+    *,
+    scale: float | None = None,
+    is_causal: bool = False,
+    implementation: Literal['xla', 'cudnn'] | None = None) -> Array:
+  r"""Scaled dot product attention function.
+
+  Computes the attention function on Query, Key, and Value tensors:
+
+  .. math ::
+    \mathrm{Attention}(Q, K, V)=\mathrm{softmax}(\frac{QK^T}{\sqrt{d_k}}V)
+
+  If we define :code:`logits` as the output of :math:`QK^T` and the
+  :code:`probs` as the output of :math:`softmax`.
+
+  Throughout this function, we utilize the following uppercase letters to
+  represent the shape of array:
+
+    B = batch size
+    S = length of the key/value (source)
+    T = length of the query (target)
+    N = number of attention heads
+    H = dimensions of each attention head
+
+  Args:
+    query: query array; shape :code:`(BTNH)`
+    key: key array; shape :code:`(BSNH)`
+    value: value array; shape :code:`(BSNH)`
+    bias: optional, bias array to be added to logits; shape broadcastable to
+      :code:`(BNTS)`.
+    mask: optional, mask array used to filter out logits. It is a boolean mask
+      where `True` indicates the element should take part in attention. For an
+      additive mask, users should pass it to `bias`. The shape is broadcastable
+      to :code:`(BNTS)`.
+    scale: scale for the logits. If None, the scale will be set to 1 divided by
+      the square root of query's head dimension (i.e. H).
+    is_causal: If true, causal attention will be applied. Note, some
+      implementations like `xla` will generate a mask tensor and apply it to the
+      logits, but other implementations like `cudnn` will avoid computing the
+      unmasked regions.
+    implementation: A string to control which implementation backend to use.
+      Supported strings are `xla`, `cudnn` (cuDNN flash attention). It defaults
+      to `None`, which will automatically select the best available backend.
+      Note, `cudnn` supports only a subset of shapes/dtypes, and an exception
+      will be thrown if its not supported.
+
+  Returns:
+    An array of the attention output with the same shape as :code:`query`.
+  """
+  def _check_has_shape(t: ArrayLike, shape: Sequence[int], name: str) -> None:
+    if t.ndim != len(shape):
+      raise ValueError(f"{name} ndim should be {len(shape)}, but got {t.ndim}")
+    value_str1 = f't.shape={t.shape}'
+    value_str2 = f'shape={shape}'
+    for i in range(t.ndim):
+      if shape[i] != -1 and t.shape[i] != shape[i]:
+        raise ValueError(f"{name} shape should be {shape}: but got {t.shape}")
+    
+  B, S, N, H = key.shape
```

Comment:
You need to call `asarray()` on `key` and other arguments before accessing `.shape`, because they are annotated to have type `ArrayLike` which e.g. includes Python scalars.

Reply:
Done.